metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/storage/LedgerStorage.scala

```diff
@@ -0,0 +1,42 @@
+package io.iohk.metronome.checkpointing.service.storage
+
+import io.iohk.metronome.checkpointing.models.Ledger
+import io.iohk.metronome.storage.{KVRingBuffer, KVCollection, KVStore}
+import scodec.Codec
+
+/** Storing the committed and executed checkpoint ledger.
+  *
+  * Strictly speaking the application only needs the committed state,
+  * since it has been signed by the federation and we know it's not
+  * going to be rolled back. Uncommitted state can be kept in memory.
+  *
+  * However we want to support other nodes catching up by:
+  * 1. requesting the latest Commit Q.C., then
+  * 2. requesting the block the Commit Q.C. points at, then
+  * 3. requesting the ledger state the header points at.
```

Comment:
Isn't that the same as latest "committed state"? What more historical state do we want to maintain?

Reply:
Do you mean consecutive committed states (ignoring the intermediate uncommited states)?

Reply:
That is indeed the latest committed state. The only need for keeping any historical state is to facilitate other members trying to catch up. If we completely forget what our state was 5 seconds ago, it might make it difficult for them to query in the window when the state is still available. 

I think we can save the state for any block that we have executed, because every executed block has carried out some effect, so ideally I don't want to execute them again. Not every block has a commit, though, the `BlockExecutor` executes them one by one, resuming from where it left off if it is restarted, since the next commit will implicitly build on any previous one. At that point ideally we should be able to restore the state that belonged to the last executed block, so it's better to save it then not. 

But uncommitted states won't be requested by other federation members to sync to, only committed ones.

Reply:
Why would a member try to sync from an old committed state? Perhaps to avoid a race condition we should keep the most recent as well as the previous committed state, but I don't see why we would need more than those 2 🤔

Reply:
For simplicity I wouldn't store the uncommitted state. Executing the blocks should be relatively cheap and we can always think about an optimisation later, though in practice I think that would be a downgrade in performance, as re-executing some blocks upon a restart every once in a while should be cheaper than storing the result of each execution.

Reply:
First let's clarify some terminology. By uncommitted state we can mean two things. 

Say we have these blocks: `B0 <- B1* <- B2 <- B3** <- B4 <- B5` and the last committed block at the moment is `B3`; before that we committed `B1`. 

By _commit_ in this instance I mean that there's an explicit Commit Quorum Certificate for those blocks. A block doesn't necessarily get a Commit Q.C. in a round, but it's also not necessarily discarded: if it has a Prepare Q.C. then it can be extended in next round. Let's say this is how `B2` and `B4` are in the block tree. 

When we validate `B5` we need to project the pre-state it's built on. Similarly if we want to extend it. These projected state are the ones I'd unequivocally call _uncommitted_ and I agree that they should not be stored. 

The other case is `B2`. This one is implicitly committed when `B3` is committed. In theory there can be an arbitrarily long sequence of blocks without a commit, but once it arrives, everything in the branch is committed in one go.

When `B3` is committed `BlockExecutor` will execute blocks `B2` and `B3`. We have bookkeeping for what the last executed block is, even if it happened to be many blocks behind the last commit. I think it makes sense to save this state, because the state change is the de-facto effect of executing a block. 

Arguing whether saving the state or re-executing blocks is cheaper only makes sense in the context of a given application. It could be really expensive and irreversible to execute a block, have real world effects that should only happen once. I think the HotStuff module should not re-execute the same block once it's been successfully done once. 

In theory the application could do its own book-keeping to what it considers the last state and try to retrieve blocks that are built on it when a restart happens, but in our case this sounds more complex than just saving the small state once every few seconds. In the case of ETC, that state won't even change with each block, since there won't be a new checkpoint at every height, so there will only be really new state once per minute.

> Why would a member try to sync from an old committed state? Perhaps to avoid a race condition we should keep the most recent as well as the previous committed state, but I don't see why we would need more than those 2

I don't think you'd deliberately try to sync from "old" committed state, but who knows how quickly something becomes old. You're exactly right, 2 should be enough to avoid the simplest race condition, but some use cases may need a bit more leeway. Maybe "state" isn't as simple as it is for us now, maybe it's a full Merle Patricia Trie that has to be negotiated for what the diff is through multiple rounds of requests.

Reply:
If there were expensive real world effects for block execution it would probably have to protect itself against re-deliveries, maybe have a nonce built into the commands, because it couldn't be sure that we correctly save the fact that a block has been executed. 

Nevertheless, another reason I thought it should be okay to save the state of blocks like `B2` is because these blocks could have their own Commit Q.C. anyway, it just so happened that they didn't, so the application should be able to handle this many writes, although maybe in a less burst-y fashion. You may be right that performance could become uneven if commits don't happen regularly, in which case we need to tune the timeout.

Reply:
Another piece of food for thought is that I added [block pruning](https://github.com/input-output-hk/metronome/pull/42/commits/7c4c9cb344ed1ac007fa7d7972279b00469454f4) to the example application based on the last executed block hash, so if the application happened to not save a lot of the state, it may not be able to reconstruct it after a restart, although this would be extreme. And in any case the checkpointing app can have different pruning strategy, one based on what it thinks the last restartable block was.